patches/tModLoader/Terraria/Main.TML.cs

```diff
@@ -170,37 +171,62 @@ public partial class Main
 			}
 		}
 
-		internal void InitTMLContentManager() {
-			if (dedServ) {
-				return;
+		public static bool IsValidContentFolder(string vanillaContentFolder, out string reason) {
+			if (!Directory.Exists(vanillaContentFolder)) {
+				reason = Language.GetTextValue("tModLoader.ContentFolderNotFound");
+				return false;
 			}
 
-			string vanillaContentFolder;
-			if (SocialAPI.Mode == SocialMode.Steam) {
-				vanillaContentFolder = Path.Combine(Steam.GetSteamTerrariaInstallDir(), "Content");
+			// Canary file, ensures that Terraria has updated to at least the version this tModLoader was built for. Alternate check to BuildID check in TerrariaSteamClient for non-Steam launches 
+			if (!File.Exists(Path.Combine(vanillaContentFolder, "Images", "Projectile_981.xnb"))) {
+				reason = Language.GetTextValue("tModLoader.TerrariaOutOfDateMessage");
+				return false;
 			}
-			else {
-				vanillaContentFolder = Platform.IsOSX ? "../Terraria/Terraria.app/Contents/Resources/Content" : "../Terraria/Content"; // Side-by-Side Manual Install
 
-				if (!Directory.Exists(vanillaContentFolder)) {
-					vanillaContentFolder = Platform.IsOSX ? "../Terraria.app/Contents/Resources/Content" : "../Content"; // Nested Manual Install
+			reason = "Valid";
+			return true;
+		}
+
+		public static string GetContentFolderPath() {
+			string vanillaContentFolder;
+			if (!File.Exists("LaunchUtils/TerrariaFolderPath.txt") || !IsValidContentFolder(File.ReadAllText("LaunchUtils/TerrariaFolderPath.txt"), out _)) {
+				if (SocialAPI.Mode == SocialMode.Steam) {
+					vanillaContentFolder = Path.Combine(Steam.GetSteamTerrariaInstallDir(), "Content");
+				}
+				else {
+					vanillaContentFolder = Platform.IsOSX ? "../Terraria/Terraria.app/Contents/Resources/Content" : "../Terraria/Content"; // Side-by-Side Manual Install
+
+					if (!Directory.Exists(vanillaContentFolder)) {
+						vanillaContentFolder = Platform.IsOSX ? "../Terraria.app/Contents/Resources/Content" : "../Content"; // Nested Manual Install
+					}
 				}
-				Logging.tML.Info("Content folder of Terraria GOG Install Location assumed to be: " + Path.GetFullPath(vanillaContentFolder));
+
+				if (!IsValidContentFolder(vanillaContentFolder, out _)) {
+					nativefiledialog.NFD_PickFolder(null, out vanillaContentFolder);
+				}
+
+				File.WriteAllText("LaunchUtils/TerrariaFolderPath.txt", vanillaContentFolder);
```

Comment:
Recommend moving this down below, as a 'common retry' code. Allows you to simplify some of the conditions here.

Also recommend only writing the `LaunchUtils/TerrariaFolderPath.txt` file if it's user selected.

Also better not to have user configuration items in the `LaunchUtils` folder. Just save this in the root directory and call it `ContentFolderPath.txt`